File: src/main/java/net/benwoodworth/fastcraft/core/api/recipe/RecipeProvider.java
package net.benwoodworth.fastcraft.core.api.recipe;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

/**
 * Provides recipe from the server.
 *
 * @param <TItem> The native item type.
 */
public abstract class RecipeProvider<TItem> {

    /**
     * Get recipe sources.
     */
    @NotNull
    abstract List<RecipeSource<TItem>> getRecipeSources();

    /**
     * Get recipe on the server.
     */
    @NotNull
    public List<FcRecipe<TItem>> getServerCraftingRecipes() {
        List<FcRecipe<TItem>> recipes = new ArrayList<>();
        getRecipeSources().forEach(source ->
                recipes.addAll(source.getCraftingRecipes())
        );
        return recipes;
    }
}

File: src/main/kotlin/net/benwoodworth/fastcraft/core/dependencies/recipes/RecipeProvider.kt
package net.benwoodworth.fastcraft.core.dependencies.recipes

/**
 * Provides recipe from the server.
 */
interface RecipeProvider {

    /**
     * Get recipe sources.
     *
     * @return the recipe sources
     */
    fun getRecipeSources(): List<RecipeSource>

    /**
     * Get the recipes provided by the native server.
     *
     * @return the server crafting recipes
     */
    fun getServerCraftingRecipes(): List<Recipe>
}
